tests/test_csv.py

```diff
@@ -0,0 +1,134 @@
+
+directive = CSVFilterDirective(
+    None, None, None, None, None, None, None, None, None)
+tmp_directive_options = {
+    'header': 'Company,Contact,Country,Attend?',
+    'delim': '\t',
+    'file': 'example.csv',
+    'included_cols': [0, 1, 2]
+}
+
+
+def test_csv_filter_exclude_directive():
+    directive_options = copy(tmp_directive_options)
+    directive_options['exclude'] = {3: '(?i)Y\\w*'}
+    directive.options = directive_options
+
+    with mock.patch.object(directive, '_apply_filters') as apply_filters:
+        rows, max_cols = directive.parse_csv_data_into_rows(*directive_params)
+        args = apply_filters.call_args
+        assert args[0][2] is None
+        assert args[0][3] == {3: re.compile('(?i)Y\\w*', re.IGNORECASE)}
+
+
+def test_csv_filter_include_directive():
+    directive_options = copy(tmp_directive_options)
+    directive_options['include'] = {3: '(?i)Y\\w*'}
+    directive.options = directive_options
+
+    with mock.patch.object(directive, '_apply_filters') as apply_filters:
+        rows, max_cols = directive.parse_csv_data_into_rows(*directive_params)
+        args = apply_filters.call_args
+        assert args[0][2] == {3: re.compile('(?i)Y\\w*', re.IGNORECASE)}
+        assert args[0][3] is None
+
+
+def test_csv_filter_include_exclude_directive():
+    directive_options = copy(tmp_directive_options)
+    directive_options['include'] = {3: '(?i)Y\\w*'}
+    directive_options['exclude'] = {3: '(?i)Y\\w*'}
+    directive.options = directive_options
+
+    with mock.patch.object(directive, '_apply_filters') as apply_filters:
+        rows, max_cols = directive.parse_csv_data_into_rows(*directive_params)
+        args = apply_filters.call_args
+        assert args[0][2] == {3: re.compile('(?i)Y\\w*', re.IGNORECASE)}
+        assert args[0][3] == {3: re.compile('(?i)Y\\w*', re.IGNORECASE)}
+
+
+rows = [
+    [
+        (0, 0, 0, StringList(['Centro comercial Moctezuma'])),
+        (0, 0, 0, StringList(['Francisco Chang'])),
+        (0, 0, 0, StringList(['Mexico'])),
+        (0, 0, 0, StringList(['YES'])),
+    ],
+    [
+        (0, 0, 0, StringList(['Alfreds Futterkiste'])),
+        (0, 0, 0, StringList(['Maria Anders'])),
+        (0, 0, 0, StringList(['Germany'])),
+        (0, 0, 0, StringList(['y'])),
+    ],
+    [
+        (0, 0, 0, StringList(['Ernst Handel'])),
+        (0, 0, 0, StringList(['Roland Mendel'])),
+        (0, 0, 0, StringList(['Austria'])),
+        (0, 0, 0, StringList(['NO'])),
+    ]
+]
+
+expected = [
+    [
+        (0, 0, 0, StringList(['Ernst Handel'])),
+        (0, 0, 0, StringList(['Roland Mendel'])),
+        (0, 0, 0, StringList(['Austria'])),
+        (0, 0, 0, StringList(['NO'])),
+    ]
+]
+
+
+def test_apply_exclude_filters():
+    exclude_filters = {3: re.compile('(?i)Y\\w*', re.IGNORECASE)}
+    result = directive._apply_filters(rows, 4, None, exclude_filters)
+    assert result == expected
+
+
+def test_apply_include_filters():
```

Comment:
I'd love to see two more tests:
- one that doesn't do filtering at all
- one that combines include and exclude, e.g. include only 2 out of the 3 rows, but then exclude 1 of the 2 remaining rows; possibly over different columns. E.g. include `{0: r"^[CE].+$"}` and exclude `{1: r"M"}`. That should leave the row with `Mexico`.